--- a/Controller/Category/View.php
+++ b/Controller/Category/View.php
@@ -152,7 +152,8 @@ protected function _initCategory()
      */
     public function execute()
     {
-        if ($this->_request->getParam(\Magento\Framework\App\ActionInterface::PARAM_NAME_URL_ENCODED)) {
+        if ($this->_request->getParam(\Magento\Framework\App\ActionInterface::PARAM_NAME_URL_ENCODED)
+            && !$this->_request->getParam('___from_store')) {
             return $this->resultRedirectFactory->create()->setUrl($this->_redirect->getRedirectUrl());
         }
         $category = $this->_initCategory();
diff --git a/app/code/Magento/Catalog/Controller/Product/View.php b/app/code/Magento/Catalog/Controller/Product/View.php
index 27def2125ca1..515634669ba7 100644
--- a/Controller/Product/View.php
+++ b/Controller/Product/View.php
@@ -76,7 +76,8 @@ public function execute()
         $productId = (int) $this->getRequest()->getParam('id');
         $specifyOptions = $this->getRequest()->getParam('options');

-        if ($this->getRequest()->isPost() && $this->getRequest()->getParam(self::PARAM_NAME_URL_ENCODED)) {
+        if ($this->getRequest()->isPost() && $this->getRequest()->getParam(self::PARAM_NAME_URL_ENCODED)
+            && !$this->_request->getParam('___from_store')) {
             $product = $this->_initProduct();
             if (!$product) {
                 return $this->noProductRedirect();