diff --git a/module-catalog/Model/Product/Option/SaveHandler.php b/module-catalog/Model/Product/Option/SaveHandler.php
index ff15f88..c1ef351 100644
--- a/module-catalog/Model/Product/Option/SaveHandler.php
+++ b/module-catalog/Model/Product/Option/SaveHandler.php
@@ -3,9 +3,10 @@
  * Copyright © Magento, Inc. All rights reserved.
  * See COPYING.txt for license details.
  */
+
 namespace Magento\Catalog\Model\Product\Option;
 
-use Magento\Catalog\Api\ProductCustomOptionRepositoryInterface as OptionRepository;
+use Magento\Catalog\Model\Product\Option\Repository as OptionRepository;
 use Magento\Framework\EntityManager\Operation\ExtensionInterface;
 
 /**
@@ -23,7 +24,8 @@ class SaveHandler implements ExtensionInterface
      */
     public function __construct(
         OptionRepository $optionRepository
-    ) {
+    )
+    {
         $this->optionRepository = $optionRepository;
     }
 
@@ -35,17 +37,27 @@ class SaveHandler implements ExtensionInterface
      */
     public function execute($entity, $arguments = [])
     {
+        $origOptions = $entity->getOrigData()['options'];
         $options = $entity->getOptions();
+
+        $map = [];
         $optionIds = [];
 
         if ($options) {
-            $optionIds = array_map(
-                function ($option) {
-                    /** @var \Magento\Catalog\Model\Product\Option $option */
-                    return $option->getOptionId();
-                },
-                $options
-            );
+            //Map already existing options
+            foreach ($origOptions as $origOption) {
+                $values = array();
+                foreach ($origOption->getValues() as $value) {
+                    $values[$value->getTitle()] = $value->getOptionTypeId();
+                }
+
+                $map[$origOption->getSku()] = [
+                    'option_id' => $origOption->getOptionId(),
+                    'values' => $values
+                ];
+
+                $optionIds[] = $origOption->getOptionId();
+            }
         }
 
         /** @var \Magento\Catalog\Api\Data\ProductInterface $entity */
@@ -54,9 +66,25 @@ class SaveHandler implements ExtensionInterface
                 $this->optionRepository->delete($option);
             }
         }
-
         if ($options) {
             foreach ($options as $option) {
+                if (is_null($option->getOptionId()) && array_key_exists($option->getSku(), $map)) {
+                    $option->setOptionId($map[$option->getSku()]['option_id']);
+                }
+
+                $optionDataValues = $option->getData('values');
+                $refactoredValues = [];
+                foreach ($optionDataValues as $val){
+                    if (is_null($val['option_type_id']) && array_key_exists($val['title'], $map[$option->getSku()]['values'])) {
+                        $val['option_type_id'] = $map[$option->getSku()]['values'][$val['title']];
+                    }
+                    $val['mageworx_option_type_id'] = null;
+                    $refactoredValues[] = $val;
+                }
+
+                $option->setProductSku($entity->getSku());
+                $option->unsetData('values');
+                $option->setData('values' , $refactoredValues);
                 $this->optionRepository->save($option);
             }
         }
