--- a/lib/web/tiny_mce/tiny_mce_src.js
+++ b/lib/web/tiny_mce/tiny_mce_src.js
@@ -1166,11 +1166,8 @@ tinymce.create('static tinymce.util.XHR', {
                ed.onClick.add(function(ed, e) {
                        e = e.target;

-                       // Workaround for bug, http://bugs.webkit.org/show_bug.cgi?id=12250
-                       // WebKit can't even do simple things like selecting an image
-                       // Needs tobe the setBaseAndExtend or it will fail to select floated images
                        if (/^(IMG|HR)$/.test(e.nodeName))
-                               ed.selection.getSel().setBaseAndExtent(e, 0, e, 1);
+                               ed.selection.select(e);

                        if (e.nodeName == 'A' && ed.dom.hasClass(e, 'mceItemAnchor'))
                                ed.selection.select(e);