diff --git a/view/adminhtml/web/js/variations/variations.js b/view/adminhtml/web/js/variations/variations.js
--- a/view/adminhtml/web/js/variations/variations.js
+++ b/view/adminhtml/web/js/variations/variations.js
@@ -410,12 +410,8 @@ define([
             this.source.data['configurable-matrix-serialized'] =
                 JSON.stringify(this.source.data['configurable-matrix']);

-            delete this.source.data['configurable-matrix'];
-
             this.source.data['associated_product_ids_serialized'] =
                 JSON.stringify(this.source.data['associated_product_ids']);
-
-            delete this.source.data['associated_product_ids'];
         },

         /**
