--- a/Model/StoreSwitcher/RewriteUrl.php
+++ b/Model/StoreSwitcher/RewriteUrl.php
@@ -65,17 +65,16 @@ class RewriteUrl implements StoreSwitcherInterface
             UrlRewrite::STORE_ID => $oldStoreId,
         ]);
         if ($oldRewrite) {
+            $targetUrl = $targetStore->getBaseUrl();
             // look for url rewrite match on the target store
             $currentRewrite = $this->urlFinder->findOneByData([
-                UrlRewrite::REQUEST_PATH => $urlPath,
+                UrlRewrite::TARGET_PATH => $oldRewrite->getTargetPath(),
                 UrlRewrite::STORE_ID => $targetStore->getId(),
             ]);
-            if (null === $currentRewrite) {
-                /** @var \Magento\Framework\App\Response\Http $response */
-                $targetUrl = $targetStore->getBaseUrl();
+            if ($currentRewrite) {
+                $targetUrl .= $currentRewrite->getRequestPath();
             }
         }
-
         return $targetUrl;
     }
 }
